algorithms/src/snark/marlin/ahp/verifier/verifier.rs

```diff
@@ -20,78 +20,114 @@ use crate::{
+        batch_sizes: &BTreeMap<CircuitId, usize>,
+        circuit_infos: &BTreeMap<&'a CircuitId, &CircuitInfo<TargetField>>,
+        max_constraint_domain: EvaluationDomain<TargetField>,
+        largest_non_zero_domain: EvaluationDomain<TargetField>,
         fs_rng: &mut R,
-    ) -> Result<(FirstMessage<TargetField>, State<TargetField, MM>), AHPError> {
-        // Check that the R1CS is a square matrix.
-        if index_info.num_constraints != index_info.num_variables {
-            return Err(AHPError::NonSquareMatrix);
-        }
-
-        let constraint_domain_time = start_timer!(|| "Constructing constraint domain");
-        let constraint_domain =
-            EvaluationDomain::new(index_info.num_constraints).ok_or(AHPError::PolynomialDegreeTooLarge)?;
-        end_timer!(constraint_domain_time);
-
-        let non_zero_a_time = start_timer!(|| "Constructing non-zero-a domain");
-        let non_zero_a_domain =
-            EvaluationDomain::new(index_info.num_non_zero_a).ok_or(AHPError::PolynomialDegreeTooLarge)?;
-        end_timer!(non_zero_a_time);
-
-        let non_zero_b_time = start_timer!(|| "Constructing non-zero-b domain");
-        let non_zero_b_domain =
-            EvaluationDomain::new(index_info.num_non_zero_b).ok_or(AHPError::PolynomialDegreeTooLarge)?;
-        end_timer!(non_zero_b_time);
-
-        let non_zero_c_time = start_timer!(|| "Constructing non-zero-c domain");
-        let non_zero_c_domain =
-            EvaluationDomain::new(index_info.num_non_zero_c).ok_or(AHPError::PolynomialDegreeTooLarge)?;
-        end_timer!(non_zero_c_time);
-
-        let input_domain_time = start_timer!(|| "Constructing input domain");
-        let input_domain =
-            EvaluationDomain::new(index_info.num_public_inputs).ok_or(AHPError::PolynomialDegreeTooLarge)?;
-        end_timer!(input_domain_time);
-
-        let squeeze_time = start_timer!(|| "Squeezing challenges");
-        let elems = fs_rng.squeeze_nonnative_field_elements(3 + batch_size - 1);
+    ) -> Result<(FirstMessage<'a, TargetField>, State<'a, TargetField, MM>), AHPError> {
+        let elems = fs_rng.squeeze_nonnative_field_elements(3);
         let (first, rest) = elems.split_at(3);
         let [alpha, eta_b, eta_c]: [_; 3] = first.try_into().unwrap();
-        let mut batch_combiners = vec![TargetField::one()];
-        batch_combiners.extend_from_slice(rest);
-        end_timer!(squeeze_time);
+        let mut batch_combiners = BTreeMap::new();
+        let mut circuit_specific_states = BTreeMap::new();
+        let mut circuit_combiners_needed = 0; // the first circuit_combiner is simply TargetField::one()
+
+        for (batch_size, (circuit_id, circuit_info)) in batch_sizes.values().zip(circuit_infos) {
+            let squeeze_time = start_timer!(|| "Squeezing challenges");
+            // TODO: we should do a review as to what happens when we have more than usize circuit/instance combiners
```

Comment:
We can just return an error; if we hit that many then the system chain will fall over lol